### DISCUSSION TITLE
2023/06/09/17.callback/

### DISCUSSION THREAD

**Discussion by giscus** [qllokirin/hexo_blog_comments#6] [Announcements]
# 2023/06/09/17.callback/

https://qllokirin.github.io/2023/06/09/17.callback/

**Comment by zzsqwq** [comment]
> 从结果来看，队列长度里面储存的内容是在上一次 callback 执行时更新的，如果callback花费很长时间，这个队列为一的消息并不是最新的，而是上一次 callback 运行时候的值。

回想起来这个问题我也遇到过，但是当时通过改了订阅者的 buff_size 解决的，但是没有深究。但是今天又重新看了一下，大概搞懂了原因，具体链接可以看 [Image subscriber lag (despite queue = 1)](https://answers.ros.org/question/220502/image-subscriber-lag-despite-queue-1/?answer=220505#post-id-220505) ，大概是说由于 buff_size 每次只能接受半帧导致的。但是现在学的多了，感觉无论什么时候的最佳实践应该都是自己维护队列在 callback 中存想要的消息，然后另开一个线程处理队列里面的消息。

**Reply by qllokirin** [reply]
buff_size 我还没有用过，我搜了一下看到下面这句

> 要设置Subscriber的buff_size（缓冲区大小）足够大，大于一个消息的大小

我试着在订阅者加入 buff_size=52428800，但是没用任何改变
看起来对于时间消耗大的程序，回调函数只负责更新消息才是最佳选择，不管怎样，算是找到一个比较河里的解决方法
对于传统视觉，处理速度能到百帧以上，远大于一般相机的30Hz，感觉这时候直接使用spin也挺不错的
